CommonLib/Error/GREYErrorFormatter.m

```diff
@@ -0,0 +1,197 @@
+static NSString *const kHierarchyUserInteractionEnabledLegendKey = @"[UIE]";
+static NSString *const kHierarchyBackWindowKey                   = @"Back-Most Window";
+static NSString *const kHierarchyAccessibilityKey                = @"Accessibility";
+static NSString *const kHierarchyUserInteractionEnabledKey       = @"User Interaction Enabled";
+static NSString *const kHierarchyLegendKey                       = @"Legend";
+static NSString *const kHierarchyHeaderKey                       = @"UI Hierarchy (ordered by wind"
+                                                                   @"ow level, back to front):\n";
+
+#pragma mark - GREYErrorFormatterFlag
+
+/**
+ * Keys used by GREYErrorFormatter to format a GREYError's userInfo properties.
+ * These states are not mutually exclusive and can be combined together using Bitwise-OR to
+ * represent multiple states.
+ * If more than 32 options exists, change the bitshifted values to UL.
+ */
+typedef NS_OPTIONS(NSUInteger, GREYErrorFormatterFlag) {
+  /** Empty Error Description  */
+  GREYErrorFormatterNone = 0,
+  /** Exception Reason */
+  GREYErrorFormatterFlagExceptionReason = 1 << 0,
+  /** Recovery Suggestion */
+  GREYErrorFormatterFlagRecoverySuggestion = 1 << 1,
+  /** Element Matcher */
+  GREYErrorFormatterFlagElementMatcher = 1 << 2,
+  /** Search API Info  */
+  GREYErrorFormatterFlagSearchActionInfo = 1 << 3,
+  /** Assertion Criteria, or Action Name */
+  GREYErrorFormatterFlagCriteria = 1 << 4,
+  /** Underlying ("Nested") Error */
+  GREYErrorFormatterFlagUnderlyingError = 1 << 5,
+  /** App UI Hierarchy */
+  GREYErrorFormatterFlagUIHierarchy = 1 << 6,
+};
+
+#pragma mark - GREYErrorFormatter
+
+@implementation GREYErrorFormatter
+
+#pragma mark - Public Methods
+
++ (NSString *)formattedDescriptionForError:(GREYError *)error {
+  if (GREYShouldUseErrorFormatterForError(error)) {
+    return loggerDescription(error);
+  }
+  return [GREYObjectFormatter formatDictionary:[error grey_descriptionDictionary]
+                                        indent:kGREYObjectFormatIndent
+                                     hideEmpty:YES
+                                      keyOrder:nil];
+}
+
+#pragma mark - Public Functions
+
+BOOL GREYShouldUseErrorFormatterForError(GREYError *error) {
+  return [error.domain isEqualToString:kGREYInteractionErrorDomain] &&
+          error.code == kGREYInteractionElementNotFoundErrorCode;
+}
+
+BOOL GREYShouldUseErrorFormatterForExceptionReason(NSString *reason) {
+  return [reason containsString:@"the desired element was not found"];
+}
+
+#pragma mark - Static Functions
+
+// The keys whose values should be supplied in the formatted error output.
+static NSUInteger loggerKeys(GREYError *error) {
+  if ([error.domain isEqualToString:kGREYInteractionErrorDomain] &&
+       error.code == kGREYInteractionElementNotFoundErrorCode) {
+     return GREYErrorFormatterFlagExceptionReason |
```

Comment:
In the other PR's this key for Exception Reason get repeated. Are all the keys below always populated irrespective of the error type? If not, in the LoggerDescription method below - can you print a log statement based on if its source is nil / not nil?

e.g.

```
if (keys & GREYErrorFormatterFlagExceptionReason) {
    NSString *exceptionReason = error.localizedDescription;
    [logger addObject:[NSString stringWithFormat:@"\n%@", exceptionReason]];
  }
```
Can this instead be - 

```
NSString *exceptionReason = error.localizedDescription;
if (exceptionReason) {
    [logger addObject:[NSString stringWithFormat:@"\n%@", exceptionReason]];
  }
```

Reply:
Good point, resolved. Also resolved for underlying error, since both of those will always be present.